thrust/system/cuda/detail/scan_by_key.h

```diff
@@ -29,758 +29,256 @@
   {
-    template <class Arch>
-    struct PtxPlan : PtxPolicy<128> {};
+    THRUST_INDEX_TYPE_DISPATCH2(status,
+                                Dispatch32::Dispatch,
+                                Dispatch64::Dispatch,
+                                num_items,
+                                (nullptr,
+                                 tmp_size,
+                                 keys_unwrap,
+                                 values_unwrap,
+                                 result_unwrap,
+                                 equality_op,
+                                 scan_op,
+                                 cub::NullType{},
+                                 num_items_fixed,
+                                 stream,
+                                 THRUST_DEBUG_SYNC_FLAG));
+    thrust::cuda_cub::throw_on_error(status,
+                                     "after determining tmp storage "
+                                     "requirements for inclusive_scan_by_key");
+  }
 
-    typedef core::specialize_plan<PtxPlan> ptx_plan;
+  // Run scan:
+  {
+    // Allocate temporary storage:
+    thrust::detail::temporary_array<thrust::detail::uint8_t, Derived> tmp{
+      policy,
+      tmp_size};
+
+    THRUST_INDEX_TYPE_DISPATCH2(status,
+                                Dispatch32::Dispatch,
+                                Dispatch64::Dispatch,
+                                num_items,
+                                (tmp.data().get(),
+                                 tmp_size,
+                                 keys_unwrap,
+                                 values_unwrap,
+                                 result_unwrap,
+                                 equality_op,
+                                 scan_op,
+                                 cub::NullType{},
+                                 num_items_fixed,
+                                 stream,
+                                 THRUST_DEBUG_SYNC_FLAG));
+
+    thrust::cuda_cub::throw_on_error(
+      status, "after dispatching inclusive_scan_by_key kernel");
+
+    thrust::cuda_cub::throw_on_error(
+      thrust::cuda_cub::synchronize_optional(policy),
+      "inclusive_scan_by_key failed to synchronize");
+  }
 
-    //---------------------------------------------------------------------
-    // Agent entry point
-    //---------------------------------------------------------------------
+  return result + num_items;
+}
 
-    THRUST_AGENT_ENTRY(ScanTileState tile_state,
-                       Size          num_tiles,
-                       char * /*shmem*/)
-    {
-      tile_state.InitializeStatus(num_tiles);
-    }
-  }; // struct InitAgent
+__thrust_exec_check_disable__
+template <typename Derived,
+          typename KeysInIt,
+          typename ValuesInIt,
+          typename ValuesOutIt,
+          typename InitValueT,
+          typename EqualityOpT,
+          typename ScanOpT,
+          typename SizeT>
+__host__ __device__
+ValuesOutIt exclusive_scan_by_key_n(
```

Comment:
Good catch, I'll clean this up to reduce duplication.

Reply:
On second thought, let's keep these separate -- otherwise we'll just have to re-split them when we address NVIDIA/cub#384 and add `InitialValue` support to inclusive scans.